File: TheMetz/Services/PullRequestService.cs
using System.Collections.Concurrent;
using System.Text.Json;
using Microsoft.TeamFoundation.SourceControl.WebApi;
using Microsoft.VisualStudio.Services.WebApi;
using TheMetz.Repositories;

namespace TheMetz.Services
{
    public interface IPullRequestService
    {
        public Task UpdateAdoPullRequests();
        public Task<List<GitPullRequest>> GetPullRequestsByDateOpened(int numberOfDaysAgoOpened);
        public Task<List<GitPullRequest>> GetPullRequestsByDateClosed(int numberOfDaysAgoClosed);
        public Task<List<GitPullRequest>> GetPullRequestsByDateOpenedOrClosed(int numberOfDaysAgoOpened, int numberOfDaysAgoClosed = 0);
    }

    internal class PullRequestService : IPullRequestService
    {
        private readonly VssConnection _connection;
        private readonly IPrRepository _prRepository;

        private readonly List<(string projectName, List<(string, Guid)> repos)> _projectInfo =
        [
            ("Marketplace", new List<(string repoName, Guid repoId)>
            {
                ("Marketplace", Guid.Empty),
                ("MarketplaceIDSConsumers", Guid.Empty),
                ("Clark Chains", Guid.Empty),
                ("Marketplace Customer Configuration Consumers", Guid.Empty),
                ("ClarkChainsDatabase", Guid.Empty),
                ("Marketplace-react-app", Guid.Empty),
                ("MarketplaceFrontends", Guid.Empty)
            }),
            ("Concorde", new List<(string repoName, Guid repoId)>
            {
                ("Concorde", Guid.Empty),
                ("Clark.Concorde.Models", Guid.Empty),
                ("ConcordeConsumers", Guid.Empty),
                ("ConcordeDatabase", Guid.Empty)
            }),
            ("Atlas", new List<(string repoName, Guid repoId)>
            {
                ("Atlas", Guid.Empty),
                ("Atlas.SalesHistoryModal", Guid.Empty),
                ("AtlasConsumers", Guid.Empty),
                ("AtlasDatabase", Guid.Empty)
            }),
            ("Specialty Channel Development", new List<(string repoName, Guid repoId)>
            {
                ("SpecChannelAirport", Guid.Empty),
                ("SpecChannelAirportDatabase", Guid.Empty),
                ("SpecChannelAirportConsumers", Guid.Empty)
            })
        ];

        private int _currentNumberOfDays = 0;
        private readonly DateTime _prCutoffDate = new DateTime(2024, 1, 1, 0, 0, 0, DateTimeKind.Utc);

        public PullRequestService(VssConnection connection, IPrRepository prRepository)
        {
            _connection = connection;
            _prRepository = prRepository;
        }

        public async Task<List<GitPullRequest>> GetPullRequestsByDateOpened(int numberOfDaysAgoOpened)
        {
            DateTime openedFromDate = DateTime.Now.AddDays(-numberOfDaysAgoOpened);
            if (openedFromDate < _prCutoffDate)
            {
                throw new ArgumentException(
                    "Cannot get pull requests before 2024-01-01 without manually changing the cutoff date.");
            }
            
            return await _prRepository.GetPullRequestsByDateOpened(openedFromDate);
        }

        public async Task<List<GitPullRequest>> GetPullRequestsByDateClosed(int numberOfDaysAgoClosed)
        {
            DateTime closedFromDate = DateTime.Now.AddDays(-numberOfDaysAgoClosed);
            if (closedFromDate < _prCutoffDate)
            {
                throw new ArgumentException(
                    "Cannot get pull requests before 2024-01-01 without manually changing the cutoff date.");
            }
            
            return await _prRepository.GetPullRequestsByDateClosed(closedFromDate);
        }

        public async Task<List<GitPullRequest>> GetPullRequestsByDateOpenedOrClosed(int numberOfDaysAgoOpened, int numberOfDaysAgoClosed = 0)
        {
            DateTime openedFromDate = DateTime.Now.AddDays(-numberOfDaysAgoOpened);
            DateTime closedFromDate = numberOfDaysAgoClosed == 0 ? openedFromDate : DateTime.Now.AddDays(-numberOfDaysAgoClosed);
            if (openedFromDate < _prCutoffDate || closedFromDate < _prCutoffDate)
            {
                throw new ArgumentException(
                    "Cannot get pull requests before 2024-01-01 without manually changing the cutoff date.");
            }


            return await _prRepository.GetPullRequestsByDateOpenedOrClosed(openedFromDate, closedFromDate);
        }

        public async Task UpdateAdoPullRequests()
        {
            (DateTime dateUpdated, GitPullRequest? gitPullRequest) latestCreatedPr = await _prRepository.GetLatestCreatedPullRequest();
            GitPullRequest? oldestOpenPr = await _prRepository.GetOldestOpenPullRequest();

            if (latestCreatedPr.gitPullRequest == null)
            {
                await StorePullRequestsFromDate(_prCutoffDate, _prCutoffDate);
            }
            else
            {
                DateTime latestResultCreationDateTime = latestCreatedPr.dateUpdated;
                DateTime oldestResultOpenDateTime = oldestOpenPr!.CreationDate;
            
                await StorePullRequestsFromDate(latestResultCreationDateTime, oldestResultOpenDateTime);
            }
        }

        private async Task LoadProjectRepos(GitHttpClient gitClient, string projectName)
        {
            (string projectName, List<(string, Guid)> repos) repoInfo =
                _projectInfo.First(p => p.projectName == projectName);

            IEnumerable<(string Name, Guid Id)> repos = (await gitClient.GetRepositoriesAsync(repoInfo.projectName))
                    .Select(r => (r.Name, r.Id))
                    .Where(r => repoInfo.repos.Any(ri =>
                    {
                        (string repoName, Guid _) = ri;
                        return repoName == r.Name;
                    })).ToList()
                ;

            repoInfo.repos = repos.ToList();
            (string projectName, List<(string, Guid)> repos) project =
                _projectInfo.First(p => p.projectName == projectName);
            _projectInfo[_projectInfo.IndexOf(project)] = repoInfo;
        }

        private async Task GetAndUpdatePullRequest(GitPullRequest pullRequest, GitHttpClient gitClient)
        {
            GitPullRequest? adoPr = await gitClient.GetPullRequestByIdAsync(pullRequest.PullRequestId);
            if (adoPr != null)
            {
                await _prRepository.UpdatePullRequestByADOPullRequestId(adoPr.PullRequestId, JsonSerializer.Serialize(adoPr));
            }
        }
        
        private async Task StorePullRequestsFromDate(DateTime prCreatedCutoffDate, DateTime prOpenCutoffDate)
        {
            using var gitClient = await _connection.GetClientAsync<GitHttpClient>();
        
            List<GitPullRequest> openPullRequests = await _prRepository.GetOpenPullRequests();
            var openPullRequestsTaskList = new List<Task>();
            foreach (GitPullRequest openPullRequest in openPullRequests)
            {
                openPullRequestsTaskList.Add(GetAndUpdatePullRequest(openPullRequest, gitClient));
            }

            await Task.WhenAll(openPullRequestsTaskList);
            
            IEnumerable<string> projectNames = _projectInfo.Select(p => p.projectName).ToList();
            foreach (string projectName in projectNames)
            {
                await LoadProjectRepos(gitClient, projectName);
            }
        
            foreach ((string projectName, List<(string name, Guid id)> repos) info in _projectInfo)
            {
                {
                    var skip = 0;
                    const int pageSize = 100;
                    List<GitPullRequest> paginatedPullRequests;
        
                    do
                    {
                        paginatedPullRequests = (await gitClient.GetPullRequestsByProjectAsync(
                            project: info.projectName,
                            searchCriteria: new GitPullRequestSearchCriteria
                                { Status = PullRequestStatus.All, IncludeLinks = true, MinTime = prOpenCutoffDate},
                            top: pageSize,
                            skip: skip
                        )).ToList();
        
                        if (paginatedPullRequests.Count != 0)
                        {
                            skip += pageSize;
        
                            await ExtractAndStoreCreatedPullRequests(prCreatedCutoffDate, paginatedPullRequests.Except(openPullRequests).ToList());
        
                        }
                    } while (paginatedPullRequests.Count == pageSize);
                }
            }
        }

        private async Task ExtractAndStoreCreatedPullRequests(DateTime prCreatedCutoffDate, List<GitPullRequest> paginatedPullRequests)
        {
            List<string> createdPrs = paginatedPullRequests
                .Where(pr =>
                    pr.CreationDate > prCreatedCutoffDate)
                .Select(pr => JsonSerializer.Serialize(pr)).ToList();

            foreach (string prJsonString in createdPrs)
            {
                await _prRepository.AddPullRequest(prJsonString);
            }
        }
    }
}

File: TheMetz.FSharp/PullRequestService.fs
namespace TheMetz.FSharp

open System
open System.Text.Json
open System.Threading.Tasks
open Microsoft.TeamFoundation.SourceControl.WebApi
open Microsoft.VisualStudio.Services.WebApi
open TheMetz.Interfaces
open FSharp.Control

type PullRequestService(connection: VssConnection, prRepository: IPrRepository, prCutoffDate: DateTime) =

    let FetchProjectRepos (repoInfo: string * (string * Guid) list) =
        task {
            let! gitClient = connection.GetClientAsync<GitHttpClient>()

            let projectName = repoInfo |> fst

            let! result = gitClient.GetRepositoriesAsync(repoInfo |> fst)

            let repos =
                result
                |> Seq.map (fun r -> r.Name, r.Id)
                |> Seq.filter (fun (name, id) -> repoInfo |> snd |> List.exists (fun (repoName, _) -> name = repoName))
                |> Seq.toList

            return (projectName, repos)
        }

    let getProjectInfoTemplate =
        [ ("Marketplace",
           [ ("Marketplace", Guid.Empty)
             ("MarketplaceIDSConsumers", Guid.Empty)
             ("Clark Chains", Guid.Empty)
             ("Marketplace Customer Configuration Consumers", Guid.Empty)
             ("ClarkChainsDatabase", Guid.Empty)
             ("Marketplace-react-app", Guid.Empty)
             ("MarketplaceFrontends", Guid.Empty) ])
          ("Concorde",
           [ ("Concorde", Guid.Empty)
             ("Clark.Concorde.Models", Guid.Empty)
             ("ConcordeConsumers", Guid.Empty)
             ("ConcordeDatabase", Guid.Empty) ])
          ("Atlas",
           [ ("Atlas", Guid.Empty)
             ("Atlas.SalesHistoryModal", Guid.Empty)
             ("AtlasConsumers", Guid.Empty)
             ("AtlasDatabase", Guid.Empty) ])
          ("Specialty Channel Development",
           [ ("SpecChannelAirport", Guid.Empty)
             ("SpecChannelAirportDatabase", Guid.Empty)
             ("SpecChannelAirportConsumers", Guid.Empty) ]) ]

    let LoadProjectInfo: Task<(string * (string * Guid) list) list> =
        let projectInfoTemplate: (string * (string * Guid) list) list =
            getProjectInfoTemplate

        task {
            let tasks = projectInfoTemplate |> List.map FetchProjectRepos

            let! projectsWithRepos = Task.WhenAll(tasks)

            return projectsWithRepos |> Array.toList
        }


    let ExtractAndStoreCreatedPullRequests prCreatedCutoffDate (paginatedPullRequests: seq<GitPullRequest>) : Task =
        task {
            let tasks =
                paginatedPullRequests
                |> Seq.filter (fun (pr: GitPullRequest) -> pr.CreationDate > prCreatedCutoffDate)
                |> Seq.map (fun pr -> JsonSerializer.Serialize(pr))
                |> Seq.map prRepository.AddPullRequest
                |> Seq.toArray

            do! Task.WhenAll(tasks)
        }

    let GetAndUpdatePullRequest (gitClient: GitHttpClient) (pullRequest: GitPullRequest) : Task =
        task {
            let! adoPr = gitClient.GetPullRequestByIdAsync(pullRequest.PullRequestId)

            if not (isNull adoPr) then
                do!
                    prRepository.UpdatePullRequestByADOPullRequestId(
                        adoPr.PullRequestId,
                        JsonSerializer.Serialize(adoPr)
                    )
        }

    let GetAndStoreADOPullRequests (gitClient: GitHttpClient) (projectName: string) prOpenCutoffDate prCreatedCutoffDate (openIds: Set<int>) =
        task {
            let pageSize = 100

            let rec loop skip =
                task {
                    let criteria =
                        GitPullRequestSearchCriteria(
                            Status = PullRequestStatus.All,
                            IncludeLinks = true,
                            MinTime = Nullable prOpenCutoffDate
                        )

                    let! page =
                        gitClient.GetPullRequestsByProjectAsync(
                            project = projectName,
                            searchCriteria = criteria,
                            top = Nullable pageSize,
                            skip = Nullable skip
                        )

                    let prs = page |> Seq.toList

                    if prs.Length > 0 then
                        let toStore = prs |> List.filter (fun pr -> not (openIds.Contains pr.PullRequestId))
                        do! ExtractAndStoreCreatedPullRequests prCreatedCutoffDate toStore

                        if prs.Length = pageSize then
                            return! loop (skip + pageSize)
                        else
                            return ()
                    else
                        return ()
                }

            do! loop 0
        }

    let GetAndUpdatePullRequest (gitClient:GitHttpClient) (pullRequest: GitPullRequest) : Task=
        task {
            let! adoPr = gitClient.GetPullRequestByIdAsync(pullRequest.PullRequestId)
            if not (isNull adoPr) then
                do! prRepository.UpdatePullRequestByADOPullRequestId(adoPr.PullRequestId, JsonSerializer.Serialize(adoPr))
        }
        
    let StorePullRequestsFromDate prCreatedCutoffDate prOpenCutoffDate : Task =
        task {
            let! gitClient = connection.GetClientAsync<GitHttpClient>()

            let! result = prRepository.GetOpenPullRequests()
            let openPullRequests = result |> Seq.toList
            let openIds =
                openPullRequests
                |> Seq.map (fun pr -> pr.PullRequestId)
                |> Set.ofSeq

            let openPullRequestsTaskList = Seq.map (GetAndUpdatePullRequest gitClient) openPullRequests
            do! Task.WhenAll(openPullRequestsTaskList)

            let! projectInfo = LoadProjectInfo
            let projectNames = projectInfo |> List.map fst

            for projectName in projectNames do
                do! GetAndStoreADOPullRequests gitClient projectName prOpenCutoffDate prCreatedCutoffDate openIds
        }

    interface IPullRequestService with
        member _.GetPullRequestsByDateOpened
            (numberOfDaysAgoOpened: int)
            : Task<Collections.Generic.List<GitPullRequest>> =
            let openedFromDate = DateTime.UtcNow.AddDays(float (-numberOfDaysAgoOpened))

            if openedFromDate < prCutoffDate then
                invalidArg
                    "numberOfDaysAgoOpened"
                    "Cannot get pull requests before 2024-01-01 without manually changing the cutoff date."

            prRepository.GetPullRequestsByDateOpened(openedFromDate)

        member _.GetPullRequestsByDateClosed
            (numberOfDaysAgoClosed: int)
            : Task<Collections.Generic.List<GitPullRequest>> =
            let closedFromDate = DateTime.UtcNow.AddDays(float (-numberOfDaysAgoClosed))

            if closedFromDate < prCutoffDate then
                invalidArg
                    "numberOfDaysAgoClosed"
                    "Cannot get pull requests before 2024-01-01 without manually changing the cutoff date."

            prRepository.GetPullRequestsByDateClosed(closedFromDate)

        member _.GetPullRequestsByDateOpenedOrClosed
            (numberOfDaysAgoOpened: int, numberOfDaysAgoClosed: int)
            : Task<Collections.Generic.List<GitPullRequest>> =
            let openedFromDate = DateTime.UtcNow.AddDays(float (-numberOfDaysAgoOpened))

            let closedFromDate =
                if numberOfDaysAgoClosed = 0 then
                    openedFromDate
                else
                    DateTime.UtcNow.AddDays(float (-numberOfDaysAgoClosed))

            if openedFromDate < prCutoffDate || closedFromDate < prCutoffDate then
                invalidArg
                    "numberOfDays"
                    "Cannot get pull requests before 2024-01-01 without manually changing the cutoff date."

            prRepository.GetPullRequestsByDateOpenedOrClosed(openedFromDate, closedFromDate)

        member _.UpdateAdoPullRequests() : Task =
            task {
            let! latestCreatedPr = prRepository.GetLatestCreatedPullRequest()
            let latestCreatedPrTuple = latestCreatedPr.ToTuple()
            let! oldestOpenPr = prRepository.GetOldestOpenPullRequest()

            if (latestCreatedPrTuple |> snd |> isNull) then
                do! StorePullRequestsFromDate prCutoffDate prCutoffDate
            else
                let latestResultCreationDateTime = latestCreatedPrTuple |> fst
                let oldestResultOpenDateTime = oldestOpenPr.CreationDate
                do! StorePullRequestsFromDate latestResultCreationDateTime oldestResultOpenDateTime
            }